FEXHeaderUtils/FEXHeaderUtils/ScopedSignalMask.h

```diff
@@ -20,27 +21,139 @@ namespace FHU {
    * Destructor Order:
    * 1) Unlock Mutex
    * 2) Unmask signals
+   *
+   * Masking signals around mutex locks is needed for signal-reentrant safety
    */
   class ScopedSignalMaskWithMutex final {
     public:
+
       ScopedSignalMaskWithMutex(std::mutex &_Mutex, uint64_t Mask = ~0ULL)
-        : Mutex {_Mutex} {
+        : Mutex {&_Mutex} {
         // Mask all signals, storing the original incoming mask
         ::syscall(SYS_rt_sigprocmask, SIG_SETMASK, &Mask, &OriginalMask, sizeof(OriginalMask));
 
         // Lock the mutex
-        Mutex.lock();
+        Mutex->lock();
+      }
+
+      // No copy or assignment possible
+      ScopedSignalMaskWithMutex(const ScopedSignalMaskWithMutex&) = delete;
+      ScopedSignalMaskWithMutex& operator=(ScopedSignalMaskWithMutex&) = delete;
+
+      // Only move
+      ScopedSignalMaskWithMutex(ScopedSignalMaskWithMutex &&rhs)
+       : OriginalMask {rhs.OriginalMask}, Mutex {rhs.Mutex} {
+        rhs.Mutex = nullptr;
       }
 
       ~ScopedSignalMaskWithMutex() {
-        // Unlock the mutex
-        Mutex.unlock();
+        if (Mutex != nullptr) {
+          // Unlock the mutex
+          Mutex->unlock();
+
+          // Unmask back to the original signal mask
+          ::syscall(SYS_rt_sigprocmask, SIG_SETMASK, &OriginalMask, nullptr, sizeof(OriginalMask));
```

Comment:
If this was moving across a thread then the signal mask of the original thread that created the mutex will have been modified then the new thread's mask would be set to the previous thread's mask.

Reply:
I'm not sure if locks can be moved across threads, and it's best to avoid for that for now, I guess.

Reply:
I'd say that we should throw a TID check in the constructor and destructor, but that would break on fork.
Should probably add some strongly worded comment on these that they should never be moved between threads.

Reply:
@Sonicadvance1 added in class description for all 3 of em